### PR TITLE
feat: implement ERC20 and ERC4626 in yvToken contract

### PR DESCRIPTION
**Pull request by jmonteer**
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Checklist

- [ ] I have run vyper and solidity linting
- [ ] I have run the tests on my machine
- [ ] I have followed commitlint guidelines
- [ ] I have rebased my changes to the latest version of the main branch

### CODE REVIEW

**Comment by anticlimactic** [comment]
not too sure why there are share balances being managed in both the yieldtoken as well as the vault contract. maybe if you could explain that, i can give better feedback.

my current feeling is that i prefer the design where there is a "dumb" erc4626 wrapper sitting on top of the vault (similar to how v2 would integrate erc4626). this maintains flexibility for integrators to either integrate via erc4626 through the wrapper or directly through the vault. having the yieldtoken manage the shares incurs extra gas and just causes function calls to bounce back and forth (i.e. extra complexity with no clear gain). also it makes more sense to have shares being issued and redeemed in the same contract that calculates them.

**Comment by jmonteer** [comment]
I agree managing balances in both contracts is inefficient. 

The current reason to do it is just to avoid implementing both ERC4626 and ERC20 at vault level. This is to make the vault code simpler and abstract it from the withdrawal logic. 

The vault is currently designed to receive the strategies where to withdraw from as an arg of the withdraw function. This could come from an off-chain UI optimiser (e.g. depending on the amount to be withdrawn it might go to a lower gas or a more liquid strategy that is more expensive to withdraw from) or from an integrator implementation. 

Another alternative that you might like is to just implement ERC4626 and ERC20 at vault level and make the ERC4626's withdrawal function (`withdraw(uint256 amount, address owner)` ) to withdraw from amount currently available in the vault (it will pass an empty array for the strategies to withdraw from. What do you think?

**Comment by anticlimactic** [comment]
I think there are three options:

- the one currently implemented here (that i do not like)
- implementing a "dumb" erc4626 wrapper (i.e. YieldToken.vy becomes VaultERC4626Wrapper and only implements 4626, with the vault housing ERC20 + vault code). this preserves the ability to allow different types of integrations as you elaborated on above (as you can either deposit directly into the vault or through the wrapper). 
- implementing ERC4626 + ERC20 in vault code (what most other ERC4626 implementations do, e.g. Solmate's implementation). I am not too sure about passing an empty `[]` as part of the strategies that you can withdraw from as this means that a user has to wait for a strategist to free funds before they can withdraw (this is not very decentralized, nor is it a very nice user experience). 

i think i still favor option 2, since it still tries to simplify the vault code while still preserving the features that we wanted from above (multiple integrations). the wrapper can also interact with a strategy registry or withdrawal queue to fetch items to withdraw from. 

am open to changing my mind if there are other alternatives/rationales

File: contracts/VaultV3.vy [hunk 1]
```diff
@@ -140,48 +143,49 @@ def erc20_safe_transfer(token: address, receiver: address, amount: uint256):
 
 # USER FACING FUNCTIONS #
 @external
-def deposit(_amount: uint256, _recipient: address) -> uint256:
+def deposit(_amount: uint256, _recipient: address, _depositor: address) -> uint256:
```
Comment (pandadefi) [inline contracts/VaultV3.vy:146]: This is not a user-facing function and will require some access control right?

File: contracts/VaultV3.vy [hunk 2]
```diff
@@ -140,48 +143,49 @@ def erc20_safe_transfer(token: address, receiver: address, amount: uint256):
...
-
-    assert amount > 0, "cannot deposit zero"
-    # TODO: should it check deposit limit?
+        amount = ASSET.balanceOf(_depositor)
 
     shares: uint256 = self._issueSharesForAmount(amount, _recipient)
+    
+    # NOTE: the caller contract is in charge of transfering funds during the callback
+    preBalance: uint256 = ASSET.balanceOf(self)
+    IYVaultDepositCallback(msg.sender).yVaultDepositCallback(amount, _depositor)
```
Comment (pandadefi) [inline contracts/VaultV3.vy:158]: Why won't you trust the callback function?

**Comment by storming0x** [comment]
> * implementing ERC4626 + ERC20 in vault code (what most other ERC4626 implementations do, e.g. Solmate's implementation). I am not too sure about passing an empty `[]` as part of the strategies that you can withdraw from as this means that a user has to wait for a strategist to free funds before they can withdraw (this is not very decentralized, nor is it a very nice user experience).
> 
I agree with @anticlimactic i think since v3 is a rewrite of v2 we can implement ERC4626 as is, and in fact a requirement of v3 IMO would be to have backward compatible interface to v2 AND erc4626 so it has the wider options for integrations, both current integrators and future ones.

Also we i think we should drop the yvShare token in favor of immutability and carry on with migrations like we do now, the down side of the yvShares and migrations seem to stack up (overhead gas, extra complexity, trust in ychad is even bigger) to overcome the benefit of adding migrations to current vault holders.

**Comment by jmonteer** [comment]
> > * implementing ERC4626 + ERC20 in vault code (what most other ERC4626 implementations do, e.g. Solmate's implementation). I am not too sure about passing an empty `[]` as part of the strategies that you can withdraw from as this means that a user has to wait for a strategist to free funds before they can withdraw (this is not very decentralized, nor is it a very nice user experience).
> 
> I agree with @anticlimactic i think since v3 is a rewrite of v2 we can implement ERC4626 as is, and in fact a requirement of v3 IMO would be to have backward compatible interface to v2 AND erc4626 so it has the wider options for integrations, both current integrators and future ones.
> 
> As shown here it shouldnt be too big of a change from v2 to add these standard methods and reuse v2 internal stuff.
> 
> https://github.com/fubuloubu/ERC4626/blob/main/contracts/VyperVault.vy
> 
> Also we i think we should drop the yvShare token in favor of immutability and carry on with migrations like we do now, the down side of the yvShares and migrations seem to stack up (overhead gas, extra complexity, trust in ychad is even bigger) to overcome the benefit of adding migrations to current vault holders.
> 
> Regarding withdrawal of locked funds, we can add additional (non standard methods) to vault to manage that specific use case.

Generally agree

Please take into account that yvShare token is not for migrations. I am 100% for immutability. This doesn't change much tho

**Comment by anticlimactic** [comment]
@jmonteer ready for review

File: contracts/VaultV3.vy [hunk 3]
```diff
@@ -354,121 +379,183 @@ def _issueSharesForAmount(amount: uint256, recipient: address) -> uint256:
...
+            assetsNeeded -= assetsToWithdraw
+
+        # if we exhaust the queue and still have insufficient total idle, revert
+        assert currTotalIdle >= assets, "insufficient total idle"
+        # commit memory to storage
+        self.totalIdle = currTotalIdle
+        self.totalDebt = currTotalDebt
+
+    self._burnShares(shares, _owner)
     self.totalIdle -= assets
```
Comment (jmonteer) [inline contracts/VaultV3.vy:451]: potential gas optimisation by updating directly in `currTotalIdle`

File: contracts/VaultV3.vy [hunk 4]
```diff
@@ -354,121 +379,183 @@ def _issueSharesForAmount(amount: uint256, recipient: address) -> uint256:
...
+        for strategy in _strategies:
+            assert self.strategies[strategy].activation != 0, "inactive strategy"
 
+            assetsToWithdraw = min(assetsNeeded, IStrategy(strategy).withdrawable())
+            # continue if nothing to withdraw
+            if assetsToWithdraw == 0:
+                continue
+
+            IStrategy(strategy).freeFunds(assetsToWithdraw)
+            ASSET.transferFrom(strategy, self, assetsToWithdraw)
```
Comment (jmonteer) [inline contracts/VaultV3.vy:433]: add 
_TODO: check that `assetsToWithdraw` have been released by using `min(ASSET.balanceOf(strategy), assetsToWithdraw)`_?

File: contracts/VaultV3.vy [hunk 5]
```diff
@@ -354,121 +379,183 @@ def _issueSharesForAmount(amount: uint256, recipient: address) -> uint256:
...
-    log Withdraw(_sender, _receiver, _owner, assets, _shares)
+    log Withdraw(_sender, _receiver, _owner, assets, shares)
 
     return assets
 
 
+@view
+@internal
+def _maxDeposit(receiver: address) -> uint256:
+    total_assets: uint256 = self._totalAssets()
```
Comment (jmonteer) [inline contracts/VaultV3.vy:462]: should we stick to camel case?

File: contracts/VaultV3.vy [hunk 6]
```diff
@@ -354,121 +379,183 @@ def _issueSharesForAmount(amount: uint256, recipient: address) -> uint256:
...
+    deposit_limit: uint256 = self.depositLimit
+    if (total_assets >= deposit_limit):
+        return 0
+    return deposit_limit - total_assets
+
+
+@view
+@internal
+def _maxRedeem(owner: address) -> uint256:
+    return min(self.balanceOf[owner], self.totalIdle)
```
Comment (jmonteer) [inline contracts/VaultV3.vy:472]: self.balanceOf[owner] will return balance of shares held by owner, self.totalIdle will return assets

we should replace `self.totalIdle` with `self._convertToShares(self.totalIdle)`

File: contracts/VaultV3.vy [hunk 7]
```diff
@@ -688,6 +775,14 @@ def setFeeManager(newFeeManager: address):
     self.feeManager = newFeeManager
     log UpdateFeeManager(newFeeManager)
 
+
+@external
+def setDepositLimit(depositLimit: uint256):
+    # TODO: permissioning
```
Comment (jmonteer) [inline contracts/VaultV3.vy:781]: this should be governance responsibility

**Pull request merged**